pallets/metaverse/src/lib.rs

```diff
@@ -244,21 +349,187 @@ pub mod pallet {
+		}
+
+		/// Lock up and stake balance of the origin account.
+		/// New stake will be applied at the beginning of the next round.
+		#[pallet::weight(100_000)]
+		pub fn stake(
+			origin: OriginFor<T>,
+			metaverse_id: MetaverseId,
+			#[pallet::compact] value: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			// Check that metaverse is registered for staking.
+			ensure!(
+				RegisteredMetaverse::<T>::contains_key(&metaverse_id),
+				Error::<T>::NotRegisteredForStaking
+			);
+
+			// Get the staking ledger or create an entry if it doesn't exist.
+			let mut staking_info = Self::staking_info(&who);
+
+			// Ensure that staker has enough balance to stake.
+			let free_balance = T::Currency::free_balance(&who).saturating_sub(T::MinStakingAmount::get());
+
+			// Remove already locked funds from the free balance
+			let available_balance = free_balance.saturating_sub(staking_info);
+			let stake_amount = value.min(available_balance);
+			ensure!(stake_amount > Zero::zero(), Error::<T>::NotEnoughBalanceToStake);
+
+			// Get the latest round staking point info or create it if metaverse hasn't been staked yet so far.
+			let current_staking_round: RoundInfo<T::BlockNumber> = Self::staking_round();
+			// Get staking info of metaverse and current round
+			let mut metaverse_stake_per_round: MetaverseStakingPoints<T::AccountId, BalanceOf<T>> =
+				Self::get_metaverse_stake_per_round(&metaverse_id, current_staking_round.current)
+					.ok_or(Error::<T>::MetaverseStakingInfoNotFound)?;
+
+			// Ensure that we can add additional staker for the metaverse.
+			ensure!(
+				metaverse_stake_per_round.stakers.len() < T::MaxNumberOfStakersPerMetaverse::get() as usize,
+				Error::<T>::MaximumAmountOfStakersPerMetaverse
+			);
+			// Increment ledger and total staker value for a metaverse.
+			staking_info = staking_info
+				.checked_add(&stake_amount)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			let individual_staker = metaverse_stake_per_round.stakers.entry(who.clone()).or_default();
+			*individual_staker = individual_staker
+				.checked_add(&stake_amount)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			ensure!(
+				*individual_staker >= T::MinStakingAmount::get(),
+				Error::<T>::MinimumStakingAmountRequired,
+			);
+
+			// Update total staked value in current round
+			MetaverseStakingSnapshots::<T>::mutate(current_staking_round.current, |may_be_staking_snapshot| {
+				if let Some(snapshot) = may_be_staking_snapshot {
+					snapshot.staked = snapshot.staked.saturating_add(stake_amount)
+				}
+			});
+
+			// Update staking info of origin
+			Self::update_staking_info(&who, staking_info);
+
+			// Update staked information for contract in current round
+			MetaverseRoundStake::<T>::insert(
+				metaverse_id.clone(),
+				current_staking_round.current,
+				metaverse_stake_per_round,
+			);
+
+			Self::deposit_event(Event::<T>::MetaverseStaked(who.clone(), metaverse_id, stake_amount));
```

Comment:
In this the event, the stake_amount could be different from the value that user passed in. Would user get confused? Shall we fail earlier?

Reply:
It is the staked amount user passed in. If they pass the value below their available balance, instead of failing the transaction, we take the available balance as staked amount.